compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -428,12 +428,24 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                     }
                 }
 
-                err.span_suggestion(
-                    call_span,
-                    &format!("try calling `{}` as a method", ident),
-                    format!("self.{}({})", path_str, args_snippet),
-                    Applicability::MachineApplicable,
-                );
+                if let Some(val) = res {
+                    self.smart_resolve_context_dependent_help(
+                        &mut err,
+                        span,
+                        source,
+                        val,
+                        &path_str,
+                        &fallback_label,
+                    );
+                } else {
+                    err.span_suggestion(
+                        call_span,
+                        &format!("try calling `{}` as a method", ident),
+                        format!("self.{}({})", path_str, args_snippet),
+                        Applicability::MachineApplicable,
+                    );
+                }
```

Comment:
Ok, this PR adds technical debt and this whole function needs to be refactored to use a table of heuristics with priorities and flags indicating whether they support combining or not.
I'm not going to ask you to do this work, but let's try to reduce the technical debt in this PR at least.

- `smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply.
- in addition to "is a call" condition `call(self)` heuristic should also use the `(Res::Def(DefKind::Struct, def_id), _) if ns == ValueNS` condition to disable itself (this is the same condition that enables the "constructor is not visible here" heuristic in `fn smart_resolve_context_dependent_help`)
- these two heuristics should "link" to each other with comments telling that they use the same condition (but with an opposite sign)

Reply:
>`smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply.

8b1cf540434bb55d9615f1806b966f78b2c66c6f does this incorrectly.

>in addition to "is a call" condition `call(self)` heuristic should also use the `res is Some((Res::Def(DefKind::Struct, def_id), _)) && ns == ValueNS` condition to disable itself (this is the same condition that enables the "constructor is not visible here" heuristic in `fn smart_resolve_context_dependent_help`)

8b1cf540434bb55d9615f1806b966f78b2c66c6f misses the `ns == ValueNS` condition.

>these two heuristics should "link" to each other with comments telling that they use the same condition (but with an opposite sign)

These comments were not added.

Reply:
>`smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply

This part wasn't addressed.

Reply:
I'm not sure how to do this as if we call `smart_resolve_context_dependent_help` at the beginning of the function, it changes the diagnostic.

Reply:
Ping @petrochenkov :)